File: apis/python_interface_helpers/stk_service_group.py
import stk_service
import stk_sg_automation
import stk_smartbeat

class stk_service_cb(stk_service_cb_class):
	def __init__(self,env,cbcls):
		stk_service_cb_class.__init__(self)
		self._cbcls = cbcls
		self._env = env
	def close(self):
		self._cbcls = None
		self._env = None
	def added(self,svcgrpid,svcid,state,svcref):
		rsvg = stk_service_group.find(svcgrpid)
		rsv = stk_service.stk_service.find(svcid)
		if rsvg != None and rsv == None:
			rsv = stk_service.stk_service(self._env,None,svcid,0,None,svcref)
		else:
			self._env.__class__.log(STK_LOG_WARNING,"service group added callback called, but missing object")
			self._env.__class__.log(STK_LOG_WARNING,"service group: " + rsvg.__class__.__name__)
			self._env.__class__.log(STK_LOG_WARNING,"service: " + rsv.__class__.__name__)
		self._cbcls.added_cb(rsvg,rsv,state)
	def removed(self,svcgrpid,svcid,state):
		rsvg = stk_service_group.find(svcgrpid)
		rsv = stk_service.stk_service.find(svcid)
		if rsvg != None and rsv != None:
			self._cbcls.removed_cb(rsvg,rsv,state)
		else:
			self._env.__class__.log(STK_LOG_WARNING,"service group removed callback called, but missing object")
			self._env.__class__.log(STK_LOG_WARNING,"service group: " + rsvg.__class__.__name__)
			self._env.__class__.log(STK_LOG_WARNING,"service: " + rsv.__class__.__name__)
	def state_chg(self,svcid,old_state,new_state):
		rsv = stk_service.stk_service.find(svcid)
		if rsv != None:
			self._cbcls.state_change_cb(rsv,old_state,new_state)
		else:
			self._env.__class__.log(STK_LOG_WARNING,"service state change callback called, but missing object")
			self._env.__class__.log(STK_LOG_WARNING,"service: " + rsv.__class__.__name__)
	def smartbeat(self,svcgrpid,svcid,smartbeat):
		try:
			rsvg = stk_service_group.find(svcgrpid)
			rsv = stk_service.stk_service.find(svcid)
			rsmb = stk_smartbeat.stk_smartbeat(smartbeat)
			# Need to convert to a smartbeat object
			if rsvg != None and rsv != None:
				self._cbcls.smartbeat_cb(rsvg,rsv,rsmb)
			else:
				self._env.__class__.log(STK_LOG_WARNING,"service group smartbeat callback called, but missing object")
				self._env.__class__.log(STK_LOG_WARNING,"service group: " + rsvg.__class__.__name__)
				self._env.__class__.log(STK_LOG_WARNING,"service: " + rsv.__class__.__name__)
		except Exception, e:
			self._env.__class__.log(STK_LOG_WARNING,str(e))

class stk_service_group:
	obj_map = {}
	def __init__(self,env,group_name,id,options):
		self._svcgrp = stk_create_service_group(env.ref(),group_name,id,options.ref());
		self.__class__.obj_map[stk_get_service_group_id(self._svcgrp)] = self
		self._env = env;
	def close(self,options = None):
		stk_destroy_service_group(self._svcgrp)
		#del self.__class__.obj_map[stk_get_service_group_id(self._svcgrp)]
	def ref(self):
		return self._svcgrp
	def name(self):
		return stk_get_service_group_name(self._svcgrp)
	def invoke(self,seq):
		stk_sg_automation.stk_sga_invoke(self._svcgrp,seq.ref())
	@classmethod
	def find(cls,svcgrpid):
		if svcgrpid in cls.obj_map:
			return cls.obj_map[svcgrpid]
	@classmethod
	def add_service_cb_option(cls,env,svcgrp_opts,appcb):
		caller = stk_service_cb_caller()
		appcb.add_callback_ref(caller)
		svccb = stk_service_cb(env,appcb)
		appcb.add_callback_map_obj(svccb)
		newopts = caller.stk_append_service_cb(svcgrp_opts.ref(),svccb.__disown__())
		svcgrp_opts.update_ref(newopts)
		return svccb
	@classmethod
	def remove_service_cb_option(cls,env,svcgrp_opts,appcb):
		caller = appcb.caller()
		svccb = appcb.map_obj()
		caller.stk_remove_service_cb(svcgrp_opts.ref())
		if svccb:
			svccb.close()

File: apis/ruby_api/stk_service_group.rb
require 'stk_service'
require 'stkservice_group'
#require 'stk_sg_automation'
#require 'stk_smartbeat'

class Stk_service_cb < Stkservice_group::Stk_service_cb_class
	def initialize(env,cbcls)
		super()
		@_cbcls = cbcls
		@_env = env
	end
	def close()
		@_cbcls = nil
		@_env = nil
	end
	def added(svcgrpid,svcid,state,svcref)
		rsvg = Stk_service_group.find(svcgrpid)
		rsv = Stk_service.stk_service.find(svcid)
		if rsvg != nil and rsv == nil
			rsv = Stk_service.stk_service(@_env,nil,svcid,0,nil,svcref)
		else
			Stk_env.log(STK_LOG_WARNING,"service group added callback called, but missing object")
			Stk_env.log(STK_LOG_WARNING,"service group: " + rsvg.class.name)
			Stk_env.log(STK_LOG_WARNING,"service: " + rsv.class.name)
		end
		@_cbcls.added_cb(rsvg,rsv,state)
	end
	def removed(svcgrpid,svcid,state)
		rsvg = Stk_service_group.find(svcgrpid)
		rsv = Stk_service.stk_service.find(svcid)
		if rsvg != nil and rsv != nil
			@_cbcls.removed_cb(rsvg,rsv,state)
		else
			Stk_env.log(STK_LOG_WARNING,"service group removed callback called, but missing object")
			Stk_env.log(STK_LOG_WARNING,"service group: " + rsvg.class.name)
			Stk_env.log(STK_LOG_WARNING,"service: " + rsv.class.name)
		end
	end
	def state_chg(svcid,old_state,new_state)
		rsv = Stk_service.stk_service.find(svcid)
		if rsv != nil
			@_cbcls.state_change_cb(rsv,old_state,new_state)
		else
			Stk_env.log(STK_LOG_WARNING,"service state change callback called, but missing object")
			Stk_env.log(STK_LOG_WARNING,"service: " + rsv.class.name)
		end
	end
	def smartbeat(svcgrpid,svcid,smartbeat)
		begin
			rsvg = Stk_service_group.find(svcgrpid)
			rsv = Stk_service.stk_service.find(svcid)
			rsmb = Stk_smartbeat.stk_smartbeat(smartbeat)
			# Need to convert to a smartbeat object
			if rsvg != nil and rsv != nil
				@_cbcls.smartbeat_cb(rsvg,rsv,rsmb)
			else
				Stk_env.log(STK_LOG_WARNING,"service group smartbeat callback called, but missing object")
				Stk_env.log(STK_LOG_WARNING,"service group: " + rsvg.class.name)
				Stk_env.log(STK_LOG_WARNING,"service: " + rsv.class.name)
			end
		rescue Exception => e
			Stk_env.log(STK_LOG_WARNING,str(e))
		end
	end
end

class Stk_service_group
	@@obj_map = {}
	def initialize(env,group_name,id,options)
		@_svcgrp = Stkservice_group::stk_create_service_group(env.ref(),group_name,id,options.ref());
		@@obj_map[Stkservice_group::stk_get_service_group_id(@_svcgrp)] = self
		@_env = env;
	end
	def close(options = nil)
		Stkservice_group::stk_destroy_service_group(@_svcgrp)
	end
	def ref()
		return @_svcgrp
	end
	def name()
		return Stkservice_group::stk_get_service_group_name(@_svcgrp)
	end
	def invoke(seq)
		stk_sg_automation.stk_sga_invoke(@_svcgrp,seq.ref())
	end
	def self.find(svcgrpid)
		if @@obj_map.include? svcgrpid
			return cls.obj_map[svcgrpid]
		end
	end
	def self.add_service_cb_option(env,svcgrp_opts,appcb)
		caller = Stkservice_group::Stk_service_cb_caller.new
		appcb.add_callback_ref(caller)
		svccb = Stk_service_cb.new(env,appcb)
		appcb.add_callback_map_obj(svccb)
		newopts = caller.stk_append_service_cb(svcgrp_opts.ref(),svccb)
		svcgrp_opts.update_ref(newopts)
		return svccb
	end
	def self.remove_service_cb_option(env,svcgrp_opts,appcb)
		caller = appcb.caller()
		svccb = appcb.map_obj()
		caller.stk_remove_service_cb(svcgrp_opts.ref())
		svccb.close() if svccb
	end
end

